src/expressions.js

```diff
@@ -1,73 +1,242 @@
+        if:   template.if,
+        then: compileTemplate(template.then),
+      };
+    }
+    return {
+      if:   template.if,
+      then: compileTemplate(template.then),
+      else: compileTemplate(template.else),
+    };
+  }
+  throw new Error('compileTemplate expects valid scope expression templates');
+};
+
+const mergeParams = (paramsA, paramsB = {}) => {
+  const params = Object.assign({}, paramsA);
+  for (const [p, T] of Object.entries(paramsB)) {
+    if (params[p] && params[p] !== T) {
+      throw new Error(`Parameter ${p} cannot be both '${T}' and '${params[p]}'`);
+    }
+    params[p] = T;
+  }
+  return params;
+};
 
 /**
- * Given a scope expression template and a set of params, return a valid
- * scope expression after rendering the template in the method described in the documentation
- * for `req.authorize()` in `src/api.js`. The third parameter is a list that will have any
- * missing parameters appended to it. Returns null only in the case that an if/then expression
- * is the template and it is false.
+ * Extract parameters from a compiled scope expression template
+ *
+ * Returns an object {'<param>': '<type>'}, where type is one of:
+ *  - `string`,
+ *  - `boolean`, or
+ *  - `array` meaning array of strings.
  */
-export const expandExpressionTemplate = (template, params, missing) => {
-  if (typeof template === 'string') {
-    return splatParams(template, params, missing);
-  } else if (_.isObject(template) && template.for && template.in && template.each) {
-    let subs = _.at(params, template.in)[0];
-    if (!subs) {
-      missing.push(template.in);
-      return null;
+const extractParams = (compiledTemplate) => {
+  const ctmpl = compiledTemplate;
+  if (ctmpl instanceof Array) {
+    return ctmpl
+      .filter((value, i) => i % 2 === 1)
+      .map(p => ({[p]: 'string'}))
+      .reduce(mergeParams, {});
+  }
+  if (ctmpl.hasOwnProperty('AllOf')) {
+    return ctmpl.AllOf.map(extractParams).reduce(mergeParams, {});
+  }
+  if (ctmpl.hasOwnProperty('AnyOf')) {
+    return ctmpl.AnyOf.map(extractParams).reduce(mergeParams, {});
+  }
+  if (ctmpl.hasOwnProperty('for')) {
+    return _.omit(mergeParams({
+      [ctmpl.in]: 'array',
+      [ctmpl.for]: 'string', // require that result from 'each' can be merged with 'for' being a string
+    }, extractParams(ctmpl.each)), ctmpl.for); // omit 'for' as it's declared here
+  }
+  if (ctmpl.hasOwnProperty('if')) {
+    return mergeParams({
+      [ctmpl.if]: 'boolean',
+    }, mergeParams(
+      extractParams(ctmpl.then),
+      ctmpl.hasOwnProperty('else') ? extractParams(ctmpl.else) : {},
+    ));
+  }
+  throw new Error('extractParams expects a compiled scope expression templates');
+};
+
+/** Render a scope expression from a compiled scope expression template and parameters */
+const render = (compiledTemplate, params) => {
+  const ctmpl = compiledTemplate;
+  if (ctmpl instanceof Array) {
+    return ctmpl.map((value, i) => {
+      if (i % 2 === 1) {
```

Comment:
All of this stuff with modulo and whatever is making me nervous. It seems like there must be a more obvious way to do this kinda stuff?

Reply:
Every second string is a parameter... I actually think it's pretty sane..

Yes, the code is unreadable, but if this is wrong tests will definitely catch it :)

Reply:
+1, if you find a better way to do this... I do care about speed of the `render` function...